WalletWasabi/WabiSabi/WabiSabiCoordinator.cs

```diff
@@ -122,45 +121,50 @@ public void BanDescendant(object? sender, Block block)
 
 	public void BanDoubleSpenders(object? sender, Transaction tx)
 	{
-		bool IsWasabiCoinjoin(uint256 txId)
+		var txId = tx.GetHash();
+		if (IsWasabiCoinJoin(txId, tx))
 		{
-			var finishedCoinJoinIds = Arena.RoundStates
-				.Select(x => x.CoinjoinState)
-				.OfType<SigningState>()
-				.Where(x => x.IsFullySigned)
-				.Select(x => x.CreateUnsignedTransaction().GetHash());
-
-			return CoinJoinIdStore.Contains(txId) || finishedCoinJoinIds.Contains(txId);
+			return;
 		}
 
-		// We don't ban our own coinjoin outputs.
-		if (IsWasabiCoinjoin(tx.GetHash()))
+		var inputOutPoints = tx.Inputs.Select(x => x.PrevOut);
+		var disruptedRounds = Arena.GetRoundsContainingOutpoints(inputOutPoints);
+
+		// No round was disrupted by the received transaction. Nothing to do here.
+		if (disruptedRounds.Length == 0)
 		{
 			return;
 		}
 
-		var outPoints = tx.Inputs.Select(x => x.PrevOut);
-
-		// Detect and punish double spending coins
-		var disrupters = Arena.RoundStates
-			.Where(r => r.Phase != Phase.Ended)
-			.SelectMany(r => r.CoinjoinState.Inputs.Select(a => (RoundId: r.Id, Coin: a)))
-			.Where(x => outPoints.Any(outpoint => outpoint == x.Coin.Outpoint))
-			.ToArray();
-
-		foreach (var (roundId, offender) in disrupters)
+		// Ban all outputs created by the received transaction because it has spent coins participating in coinjoin rounds.
+		foreach (var indexedOutput in tx.Outputs.AsIndexedOutputs())
 		{
-			Warden.Prison.DoubleSpent(offender.Outpoint, offender.Amount, roundId);
+			Warden.Prison.DoubleSpent(new OutPoint(tx, indexedOutput.N), indexedOutput.TxOut.Value, disruptedRounds);
 		}
 
 		// Abort disrupted rounds
-		var disruptedRounds = disrupters.Select(x => x.RoundId).Distinct();
 		foreach (var roundId in disruptedRounds)
 		{
 			Arena.AbortRound(roundId);
 		}
 	}
 
+	private bool IsWasabiCoinJoin(uint256 txId, Transaction tx) =>
+		CoinJoinIdStore.Contains(txId) || IsFinishedCoinJoin(txId) || IsWasabiCoinJoinLookingTx(tx);
+
+	private bool IsFinishedCoinJoin(uint256 txId) =>
+		Arena.RoundStates
```

Comment:
No because if it is in signing state and matches the txid then it doesn't matter whether it already switched to Ended or not, it is one of our coinjoins.
In the first versions i tested the Phase but here we use RoundStates which is not updated immediately but after all the steps in the round have happen so, it only added problem